llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,230 @@
+    Type *Int8Ty = Type::getInt8Ty(M->getContext());
+    CASLiteralType = PointerType::get(Int8Ty, ConstantAddrspaceID);
+  }
+  ~AddrspaceReplacer() {
+    for (Function *F : FunctionsToDrop)
+      F->eraseFromParent();
+  }
+  size_t getReplacedCallsCount() { return ReplacedCallsCount; }
+
+  void runOnFunctionDeclaration(Function *F);
+
+private:
+  Module *M;
+  PointerType *CASLiteralType;
+  size_t ReplacedCallsCount = 0;
+  /// If the variadic version gets picked during FE compilation, we'll only have
+  /// 1 function to replace. However, unique declarations are emitted for each
+  /// of the non-variadic (variadic template) calls.
+  SmallVector<Function *, 8> FunctionsToDrop;
+
+  Function *getCASPrintfFunction(Function *GenericASPrintfFunc);
+  Constant *getCASLiteral(GlobalVariable *GenericASLiteral);
+};
+} // namespace
+
+char SYCLMutatePrintfAddrspaceLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLMutatePrintfAddrspaceLegacyPass,
+                "SYCLMutatePrintfAddrspace",
+                "Move SYCL printf literal arguments to constant address space",
+                false, false)
+
+// Public interface to the SYCLMutatePrintfAddrspacePass.
+ModulePass *llvm::createSYCLMutatePrintfAddrspaceLegacyPass() {
+  return new SYCLMutatePrintfAddrspaceLegacyPass();
+}
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  AddrspaceReplacer ASReplacer(&M);
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
+    if (!F.getName().startswith("_Z18__spirv_ocl_printf"))
+      continue;
+    ASReplacer.runOnFunctionDeclaration(&F);
+  }
+  return ASReplacer.getReplacedCallsCount() ? PreservedAnalyses::all()
+                                            : PreservedAnalyses::none();
+}
+
+/// Helper implementations
+namespace {
+
+/// Encapsulates the update of CallInst's literal argument.
+struct CallReplacer {
+  CallReplacer(CallInst *CI, Constant *CASArg) : CI(CI), CASArg(CASArg) {}
+  void replaceWithFunction(Function *CASPrintf) {
+    CI->setCalledFunction(CASPrintf);
+    auto *Const = CASArg;
+    // In case there's a misalignment between the updated function type and
+    // the constant literal type, create a constant pointer cast so as to
+    // duck module verifier complaints.
+    Type *ParamType = CASPrintf->getFunctionType()->getParamType(0);
+    if (Const->getType() != ParamType)
+      Const = ConstantExpr::getPointerCast(Const, ParamType);
+    CI->setArgOperand(0, Const);
+  }
+
+private:
+  CallInst *CI;
+  Constant *CASArg;
+};
+
+/// The function's effect is similar to V->stripPointerCastsAndAliases(), but
+/// also strips load/store aliases.
+Value *stripToMemorySource(Value *V) {
+  Value *MemoryAccess = V;
+  if (auto *LI = dyn_cast<LoadInst>(MemoryAccess)) {
+    Value *LoadSource = LI->getPointerOperand();
+    auto *Store = cast<StoreInst>(*llvm::find_if(
+        LoadSource->users(), [](User *U) { return isa<StoreInst>(U); }));
```

Comment:
> Technically, there could be several stores. Plus, there could be a complex CFG, where in different paths we store different values into a Value.

I would imagine that for a compile-time known literal (i.e. the only one that our pass can physically move into the constant address space), `load`/`store` combinations would lead to the same global in the end. However, I agree that multiple `load`/`store` pairs without changes to the string content could be the case. I'll see if there's a way to account for that in this helper or to diagnose the non-manageable generic AS literal instead (and update the function's description accordingly).

Reply:
Updated the function's description & started diagnosing its result more carefully at the client side.